chore: upgrade docusaurus to 2.3.1 & react 18.2

blakef opened the pull request:
Upgrade these dependencies, so we can support deep-linking in [Tabs](https://deploy-preview-8225--docusaurus-2.netlify.app/docs/markdown-features/tabs/?current-os=ios#query-string).
<!--
Thank you for the PR! Contributors like you keep React Native awesome!

Please see the Contribution Guide for guidelines:

https://github.com/facebook/react-native-website/blob/master/CONTRIBUTING.md

If your PR references an existing issue, please add the issue number below:

#<Issue>
-->

Simek commented:
Thanks for looking at Docusaurus upgrade! 👍 

There are several thing which need to be done,, beside bumping the version in `package.json`:
* we need to regenerate swizzled components, and re-apply our custom modifications, files are located in https://github.com/facebook/react-native-website/tree/main/website/src/theme, and there is a guide how to do this - https://docusaurus.io/docs/next/swizzling
* we need to verify if custom classes used in our SCSS files still exist, there is no other way than going through various pages and looking for regressions in comparison to the currently deployed PROD instance

Broken thing I have spotted on quick check:
* loading showcase page crashes Docusuarus - https://deploy-preview-3618--react-native.netlify.app/showcase
* announcement bar (at top) looks unstyled now

blakef commented:
## Push for the Showcase issue coming soon.

I wrote some visual diff tooling to give me a better chance of catching these things.
<img width="1466" alt="CleanShot 2023-03-13 at 20 47 06@2x" src="https://user-images.githubusercontent.com/49578/224828594-38929588-7a1e-4562-8348-2bcca53b69b0.png">

blakef commented:
@Simek I think this LGTM now.  When you have a moment would you mind looking over it again.  I've incorporated the swizzle changes and fixed the issue with showcase.

Simek commented on website/src/pages/showcase.js:
```diff
@@ -77,13 +77,13 @@ const Showcase = () => {
-    setRandomizedApps(randomizeApps(others.filter(app => !app.pinned)));
-    setPinnedRandomizedApps(randomizeApps(others.filter(app => app.pinned)));
-  }, []);
+  const [pinnedRandomizedApps] = useState(
+    randomizeApps(others.filter(app => !app.pinned))
+  );
+  const [randomizedApps] = useState(
+    randomizeApps(others.filter(app => app.pinned))
+  );
```
Removing `useEffect` wrap for entries randomization, leads to the rendering issue, where app labels do not match descriptions. This was written that way to address those issues.

 
<img width="1016" alt="Screenshot 2023-03-14 090331" src="https://user-images.githubusercontent.com/719641/224935113-410b88d0-1c52-4947-87f8-ca2a1509cf54.png">

Simek commented:
Thank you for all the work put into the upgrade! 🙏 Besides the comment about Showcase issue, this LGTM! 👍 

Also, it looks like we have broken content container on versions page - https://deploy-preview-3618--react-native.netlify.app/versions - but surprisingly, this already seems to be an issue for PROD deployment, so we can address this in a separate PR.

blakef commented:
I'm spamming netifly atm because the issue doesn't reproduce on my locally hosted version.  Apologies for the noise.

blakef commented on website/src/pages/showcase.js:
```diff
@@ -14,13 +14,15 @@ import ThemedImage from '@theme/ThemedImage';
+const renderApp = (app, i) => <AppBox app={app} key={`app-${app.name}-${i}`} />;
+
+const AppBox = ({app}) => {
+  const imgSource = useBaseUrl(
```
This was necessary to keep the # of hook calls consistent.  [`useBaseUrl` handles FQDNs](https://github.com/facebook/docusaurus/blob/2f75979bc5b4fd971225098e1c122efdd71d6e6f/packages/docusaurus/src/client/exports/__tests__/useBaseUrl.test.tsx#L41).
